### PR TITLE
o/snapstate: prevent components in the seed from being removed

### PR DESCRIPTION
**Pull request by natibek**
Prevent the removal of  components present in the current seed while seed-refresh is enabled.

Tracked with: [SNAPDENG-37111](https://warthogs.atlassian.net/browse/SNAPDENG-37111?atlOrigin=eyJpIjoiYzEwMzM5MmU0M2U4NDJjOWE3ZmU1NGY2ZDNmNGUzNzAiLCJwIjoiaiJ9)

[SNAPDENG-37111]: https://warthogs.atlassian.net/browse/SNAPDENG-37111?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: overlord/snapstate/component.go [hunk 1]
```diff
@@ -796,27 +796,58 @@ func RemoveComponents(st *state.State, snapName string, compName []string, opts
 	return tss, nil
 }
 
-func removeComponentTasks(st *state.State, snapst *SnapState, compst *sequence.ComponentState, info *snap.Info, setupSecurity *state.Task, copts ConflictOptions) (*state.TaskSet, error) {
+func canRemoveComponent(st *state.State, compst *sequence.ComponentState, info *snap.Info, copts ConflictOptions) error {
```
Comment (olivercalder) [inline overlord/snapstate/component.go:799]: So `snapst` was just unused previously, and is now removed as parameter?

Comment (olivercalder) [reply overlord/snapstate/component.go:799]: Ah I see, the part where `snapst` is relevant was moved to a separate `removeComponentTasks` with this function just being a check. This is nice and elegant, good work

**Review by olivercalder — APPROVED** [review]
Nice, thank you! Seems like an elegant fix to let the caller specify the component if it's relevant.

I'm no components expert so please get reviews from Andrew in particular, but LGTM.

File: overlord/snapstate/component.go [hunk 2]
```diff
@@ -796,27 +796,58 @@ func RemoveComponents(st *state.State, snapName string, compName []string, opts
...
+	deviceCtx, err := DeviceCtxFromState(st, nil)
```
Comment (andrewphelpsj) [inline overlord/snapstate/component.go]: Only define this if we need it, you can gate it behind `seedRefresh`.

File: overlord/devicestate/devicestate.go
Comment (andrewphelpsj) [inline overlord/devicestate/devicestate.go]:
```suggestion
		return errors.New("cannot remove snaps or components present in the current seed while seed-refresh is enabled")
```

File: overlord/snapstate/component.go [hunk 3]
```diff
@@ -796,27 +796,58 @@ func RemoveComponents(st *state.State, snapName string, compName []string, opts
...
-		return nil, fmt.Errorf("cannot remove component %q as it is required by an enforcing validation set", compst.SideInfo.Component)
+		return fmt.Errorf("cannot remove component %q as it is required by an enforcing validation set", compst.SideInfo.Component)
+	}
+
+	seedRefresh, err := seedRefreshEnabled(st)
+	if err != nil {
+		return err
+	}
+	if seedRefresh {
+		// Construct a component exclusive candidate with the component
```
Comment (andrewphelpsj) [inline overlord/snapstate/component.go:830]: Yeah this isn't ideal, but thanks for the comment.

**Review by andrewphelpsj — APPROVED** [review]
Thanks!

File: overlord/snapstate/component.go [hunk 4]
```diff
@@ -796,27 +796,58 @@ func RemoveComponents(st *state.State, snapName string, compName []string, opts
...
+			return err
+		}
+		// Construct a component exclusive candidate with the component
+		// to test if the component would trigger a seed refresh.
+		// The task id is not involved in the filtering when checking
+		// if a component triggers a seed, so it is fine to leave
+		// the value in the ComponentSetupTaskIDs field empty.
+		candidate := SeedRefreshCandidate{
+			InstanceName:          info.InstanceName(),
+			ComponentSetupTaskIDs: map[string]string{compst.SideInfo.Component.ComponentName: ""},
```
Comment (bboozzoo) [inline overlord/snapstate/component.go:837]: we rely on this behavior of CheckSeedRefreshRemove, and so do we need a unit test in devicestate to encode that empty task IDs doesn't matter in this case?

Comment (natibek) [reply overlord/snapstate/component.go:837]: yes, that makes sense. I added those unit tests.

**Review by bboozzoo — APPROVED** [review]
Thanks!
